File: cpp/HttpTransport.cpp
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <sstream>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <strings.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/select.h>
#include <sys/time.h>
#include <sys/uio.h>
#include "rop.h"
#include "HttpTransport.h"
#include "HttpReactor.h"
#include "base64.h"
#include "Log.h"
#include "ScopedLock.h"

using namespace std;
using namespace acba;
using namespace rop;

static const string createSessionId ()
{
    static const int LEN = 32; // create 32 byte string.
    static const char digits[] =
            "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    char s[LEN+1];
    {
        timeval t;
        gettimeofday(&t, 0);
        srand(t.tv_usec);
    }
    for (int i = 0; i < LEN; i++) {
        s[i] = digits[rand() % (sizeof(digits) - 1)];
    }
    s[LEN] = 0;
    return s;
}

void HttpConnection::readRequest ()
{
    string sid;
    HttpTransport *trans = 0;

    // if this connection is a websocket,
    if (websocketVersion) {
        if (requestPath.compare(0, 2, "/r") != 0) {
            log.warn("Unsupported websocket path:%s\n", requestPath.c_str());
            abort();
            return;
        }

        // (can be moved to other place if necessary)
        server->closeStaleTransports();

        if (requestPath.length() > 2 && requestPath[2] == '/') {
            // use already existing transport
            sid = requestPath.substr(3);
            map<string, HttpTransport*>::iterator ti =
                    server->transports.find(sid);
            if (ti != server->transports.end()) {
                trans = ti->second;
            }
            if (!trans || trans->hasWebsocket()) {
                if (trans) {
                    log.warn("Duplicated websocket connection\n");
                } else {
                    log.warn("No transport found for the websocket\n");
                }
                abort();
                return;
            }
            assert(trans->sessionId == sid);
        } else {
            // create websocket-initiated transport
            do {
                sid = createSessionId();
            } while (server->transports.find(sid) !=
                    server->transports.end());

            trans = new HttpTransport(sid, server->isRopMultiThreaded);
            if (server->exportables.size() > 0) {
                trans->registry->setExportables(server->exportables);
            }

            log.info("Created websocket-initiated transport for sid:%s\n",
                    sid.c_str());
            server->transports[sid] = trans;
        }

        // send session id
        log.debug("Sending session id...\n");
        Buffer buf;
        buf.write(sid);
        sendFrame(buf.begin(), buf.size());

        transport = trans;
        trans->setWebsocket(this);
        return;
    }

    if (requestPath.compare(0, 2, "/r") != 0) {
        log.warn("Unsupported path:%s\n", requestPath.c_str());
        sendResponse(404, "text/plain", "Unsupported xhr path");
        return;
    }

    // Create transport ahead of websocket connection
    if (requestPath.length() == 2) {
        do {
            sid = createSessionId();
        } while (server->transports.find(sid) !=
                server->transports.end());
        trans = new HttpTransport(sid, server->isRopMultiThreaded);
        if (server->exportables.size() > 0) {
            trans->registry->setExportables(server->exportables);
        }

        log.info("Created xhr-initiated transport for sid:%s\n",
                sid.c_str());
        server->transports[sid] = trans;

        sendResponse(200, "text/plain", sid);
        return;
    }

    // otherwise, treat as rop over xhr.
    sid = requestPath.substr(3);
    map<string, HttpTransport*>::iterator i = server->transports.find(sid);
    if (i == server->transports.end()) {
        log.warn("No corresponding transport for sid:%s\n", sid.c_str());
        sendResponse(404, "text/plain", "No transport here!");
        return;
    }

    i->second->touchTimeout();
    i->second->handleMessage(contentBuffer.begin(), contentBuffer.size(), this);
}

void HttpConnection::readFrame ()
{
    transport->handleMessage(contentBuffer.begin(), contentBuffer.size(), 0);
}

void HttpConnection::close ()
{
    // Delegate close request to the bound transport.
    // it will call this close() again.
    if (transport) {
        string &sid = transport->sessionId;
        log.info("closing transport for sid:%s\n", sid.c_str());
        server->transports.erase(sid);
        HttpTransport *ht = transport;
        transport = 0;
        ht->close();
        return;
    }

    log.info("closing http connection\n");
    HttpReactor::close();
    deref();
    return;

}

int HttpServer::start (int port, EventDriver *ed)
{
    struct sockaddr_in servAddr;
    bzero(&servAddr, sizeof(servAddr));
    servAddr.sin_family = AF_INET;
    servAddr.sin_addr.s_addr = htonl(INADDR_ANY);
    servAddr.sin_port = htons(port);

    int fd = socket(AF_INET, SOCK_STREAM, 0);
    {
        int optval = 1;
        setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &optval,
                sizeof(optval));
    }
    if (bind(fd, (struct sockaddr*)&servAddr, sizeof(servAddr))) {
        log.error("bind failed ret:%d\n", errno);
        ::close(fd);
        return -1;
    }
    listen(fd, 1);
    ed->watchFile(fd, acba::EVENT_READ, this);
    log.info("listening on port %d (fd:%d)\n", port, fd);
    return 0;
}

void HttpServer::stop ()
{
    abort("webserver stopped by request");
}

void HttpServer::read ()
{
    int fd;
    {
        struct sockaddr_in clientAddr;
        socklen_t len = sizeof(clientAddr);
        fd = accept(getFd(), (struct sockaddr*)&clientAddr, &len);
    }
    if (!fd) {
        log.warn("Accept failed.\n");
        return;
    }
    log.info("new connection fd:%d\n", fd);

    HttpConnection *hc = new HttpConnection(this);
    getEventDriver()->watchFile(fd, acba::EVENT_READ, hc);
}

void HttpServer::closeStaleTransports () {
    map<string, HttpTransport*>::iterator i = transports.begin();
    while (i != transports.end()) {
        if (!i->second->isStale()) {
            i++;
            continue;
        }
        log.info("Deleting stale transport:%s\n", i->first.c_str());
        i->second->close();
        transports.erase(i++);
    }
}

int HttpTransport::send (Port *p, Buffer &buf)
{
    log.debug("send message(%d bytes) to port:%d\n", buf.size(), p->id);

    if (closed) {
        log.info("sending aborted. transport was closed.\n");
        return -1;
    }

    // add port id as prefix
    int8_t h = buf.peek(0);
    buf.preWriteI32(p->id);

    XhrSession *sess = getXhrSession(p->id);
    if (sess) {
        // trace call depth.
        // it becomes 0 when no subsequent xhrs are expected.
        switch ((h >> 6) & 3) {
        case Port::MESSAGE_SYNC_CALL:
            sess->callDepth++;
            break;
        case Port::MESSAGE_RETURN:
            sess->callDepth--;
            break;
        }

        // construct send buffer
        Buffer sb;
        int len = (buf.size() + 2) / 3 * 4;

        stringstream ss;
        ss << "HTTP/1.1 200 OK\r\n"
            "Access-Control-Allow-Origin: *\r\n"
            "Content-Type: text/plain\r\n"
            "Content-Length: " << len << "\r\n"
            "\r\n";
        sb.write(ss.str());
        sb.ensureMargin(len);
        base64_encode(buf.begin(), reinterpret_cast<char*>(sb.end()),
                buf.size());
        sb.grow(len);

        if (sess == currentXhrSession && sess->responses.size() == 0) {
            // directly send the response
            assert(xmlHttpRequest);
            xmlHttpRequest->send(sb.begin(), sb.size());
            xmlHttpRequest = 0;
            currentXhrSession = 0;

            // close xhr session
            if (sess->callDepth == 0) {
                xhrSessions.remove(sess);
                delete sess;
            }

        } else {
            sess->responses.writeI32(sb.size());
            sess->responses.writeBytes(sb.begin(), sb.size());
        }
    } else { // via websocket
        if (websocket) {
            websocket->sendFrame(buf.begin(), buf.size());
        } else {
            pendingMessages.writeI32(buf.size());
            pendingMessages.writeBytes(buf.begin(), buf.size());
        }
    }
    buf.clear();
    return 0;
}

int HttpTransport::wait (Port *p)
{
    if (closed) return -1;
    if (!registry) return -1;
    if (websocket && websocket->getEventDriver()->isEventThread()) {
        ScopedUnlock ul(registry->monitor);
        websocket->getEventDriver()->wait();
    } else {
        registry->wait(p->updated);
    }
    return 0;
}

void HttpTransport::close ()
{
    acba::Ref<Registry> reg = registry; // copy ref to avoid early destroy
    if (!reg) return;

    {
        ScopedLock sl(reg->monitor);

        assert(!closed);
        closed = true;

        if (websocket) {
            websocket->abort("bound transport was closed");
            websocket = 0;
        }

        if (xmlHttpRequest) {
            xmlHttpRequest->abort("bound transport was closed");
            xmlHttpRequest = 0;
        }

        Transport::close();
    }
    reg = 0;
}

HttpTransport::XhrSession *HttpTransport::getXhrSession (int portId)
{
    for (XhrSession *sess = xhrSessions.first(); sess;
            sess = xhrSessions.next(sess)) {
        if (sess->portId == portId) {
            return sess;
        }
    }
    return 0;
}

void HttpTransport::sendNextResponse (HttpConnection *xhr,
        HttpTransport::XhrSession *sess)
{
    Buffer &res = sess->responses;
    if (!res.size()) {
        // keep xhr
        xmlHttpRequest = xhr;
        currentXhrSession = sess;
        return;
    }

    int sz = res.readI32();
    xhr->send(res.begin(), sz);
    res.drop(sz);

    if (sess->callDepth == 0 && res.size() == 0) {
        xhrSessions.remove(sess);
        delete sess;
    }
}

// this function doesn't call registry.recv() in order to handle xhr.
void HttpTransport::handleMessage (uint8_t *msg, int len, HttpConnection *xhr)
{
    ScopedLock sl(registry->monitor);

    if (len < 4) {
        log.error("Message truncated\n");
        close();
        return;
    }
    Buffer buf(msg, len);
    buf.appData = registry;
    int pid = -buf.readI32();
    XhrSession *sess = 0;

    if (xhr) {
        if (xmlHttpRequest) {
            xhr->abort("Concurrent XHR detected.");
            return;
        }

        sess = getXhrSession(pid);
        if (!buf.size()) {
            if (!sess) {
                log.error("Expected valid XHR session\n");
                close();
                return;
            }
            sendNextResponse(xhr, sess);
            return;
        }
    }

    if (buf.size() < 1) {
        log.error("Too short message");
        close();
        return;
    }
    Port *p = registry->getPort(pid);
    int8_t h = buf.readI8();

    log.info("Reading from port:%d len:%d head:%d\n", pid, (len-4), h);

    // establish an xhr session if not found.
    if (xhr && !sess) {
        sess = new XhrSession();
        sess->portId = pid;
        xhrSessions.addBack(sess);
    }

    switch ((h >> 6) & 3) {
    case Port::MESSAGE_RETURN:
        {
            RemoteCall *rc = p->removeRemoteCall();
            if (!rc) {
                log.error("Expected return wait\n");
                close();
                return;
            }
            rc->readReturn(h, buf);
            if (buf.size()) {
                log.warn("Return message has trailing %d bytes.\n", buf.size());
            }
            buf.clear();

            rc->gotReturn = true;
            pthread_cond_signal(&p->updated);
            log.debug("got return of RemoteCall:%08x index:%d.\n", rc,
                    rc->returnIndex);

            if (xhr) {
                sess->callDepth--;
                // browser sent a return, which implies that
                // initial xhr request was not yet finished.
                if (sess->callDepth <= 0) {
                    log.error("XHR call depth mismatch\n");
                    close();
                    return;
                }
            }
        }
        break;

    case Port::MESSAGE_SYNC_CALL:
        if (xhr) {
            sess->callDepth++;
        }
        // fall through

    case Port::MESSAGE_ASYNC_CALL:
        {
            int32_t oid = -buf.readI32();
            SkeletonBase *skel = registry->getSkeleton(oid);
            if (!skel) {
                log.error("No skeleton object found. id:%d\n", oid);
                close();
                return;
            }

            LocalCall *lc = new LocalCall(h, skel, *registry);
            skel->readRequest(*lc, buf);
            if (buf.size()) {
                log.warn("Request message has trailing %d bytes.\n",
                        buf.size());
            }
            buf.clear();

            log.debug("got a request of LocalCall:%08x oid:%d h:%d port:%d\n",
                    lc, skel->id, h, p->id);
            p->addLocalCall(*lc);
        }
        break;

    case Port::MESSAGE_CHAINED_CALL:
        assert(!"Not Supported");
        break;
    }

    // close xhr request if complete.
    if (xhr) {
        // send dummy response, and close it.
        if (sess->callDepth == 0) {
            xhrSessions.remove(sess);
            delete sess;
            xhr->send("HTTP/1.1 200 OK\r\n"
                "Access-Control-Allow-Origin: *\r\n"
                "Content-Type: text/plain\r\n"
                "Content-Length: 0\r\n\r\n");
            return;
        }
        sendNextResponse(xhr, sess);
    }
}


void HttpTransport::setWebsocket (HttpConnection *ws)
{
    assert(!hasWebsocket());
    websocket = ws;

    int c = 0;
    Buffer &buf = pendingMessages;
    while (buf.size() > 0) {
        int s = buf.readI32();
        websocket->sendFrame(buf.begin(), s);
        buf.drop(s);
        c++;
    }

    log.info("Sent %d pending messages.\n", c);
}

File: java/com/alticast/rop/HttpTransport.java
package com.alticast.rop;

import java.io.*;
import java.util.*;
import com.alticast.io.*;

public class HttpTransport extends Transport implements HttpRequestHandler,
        LogConsts {

    // Represents the binding of port and xhr session.
    // xhr session may span to multiple xhr connections.
    private static class XhrSession {
        public int portId; // port id attached with this xhr session.
        public int callDepth; // increase on sync call, decrease on return
        public Buffer response; // pending xhr response
        public Buffer responseTail;

        public Buffer nextResponse () {
            Buffer r = response;
            response = r.next2; // tail is ignored if next is null.
            return r;
        }

        public void addResponse (Buffer sb) {
            if (response == null) {
                response = responseTail = sb;
            } else {
                responseTail.next2 = sb;
                responseTail = sb;
            }
        }
    }

    private boolean closed = false;
    private HttpReactor webSocket;

    // xmlHttpRequest was located out of XhrSession since rop uses blocked xhr
    // and only one xhr connection may exist at a time.
    private HttpReactor xmlHttpRequest;
    private XhrSession currentXhrSession;
    private List pendingMessages;

    // list of xhr sessions.
    // Multiple xhr sessions may exist at a time as following scenario:
    // 1. browser sends *sync* call (from port A)
    // 2. webmw sends back inner *async* call
    // 3. browser sends sync call again in there. (from port B, not A)
    private List xhrSessions = new ArrayList();

    private HttpServer httpServer;
    private String sessionId;

    public HttpTransport (String sid, HttpServer srv) {
        sessionId = sid;
        httpServer = srv;
        touchTimeout();
    }

    public void setWebSocket (HttpReactor ws) {
        if (webSocket != null) {
            throw new IllegalStateException("websocket already exist");
        }
        webSocket = ws;

        try {
            // flush pending messages
            synchronized (registry) {
                if (pendingMessages == null) return;
                Iterator i = pendingMessages.iterator();
                pendingMessages = null;
                while (i.hasNext()) {
                    ws.sendMessage((Buffer)i.next());
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
            close();
        }
    }

    public boolean hasWebSocket () {
        return webSocket != null;
    }

    private XhrSession getXhrSession (int portId) {
        for (int i = 0; i < xhrSessions.size(); i++) {
            XhrSession sess = (XhrSession)xhrSessions.get(i);
            if (sess.portId == portId) {
                return sess;
            }
        }
        return null;
    }

    // send a pending response, or keep xhr for later use.
    private void sendNextResponse (HttpReactor xhr, XhrSession sess) {
        if (sess.response == null) {
            // keep xhr
            xmlHttpRequest = xhr;
            currentXhrSession = sess;
            return;
        }

        try {
            xhr.send(sess.nextResponse());
            xhr.endResponse();
        } catch (IOException e) {
            xhr.close(e);
            close();
        }

        if (sess.callDepth == 0 && sess.response == null) {
            xhrSessions.remove(sess);
        }
    }

    // this function doesn't call registry.recv() in order to handle xhr.
    private void handleMessage (Buffer msg, HttpReactor xhr) {
        //msg.dump();

        // read pid first.
        if (msg.size() < 4) {
            throw new RemoteException("Empty message");
        }
        int pid = -msg.readI32();
        XhrSession sess = null;

        synchronized (registry) {

            // xhr specific
            if (xhr != null) {

                // previous connection must be null
                if (xmlHttpRequest != null) {
                    xhr.close("Concurrent XHR detected.");
                    return;
                }

                sess = getXhrSession(pid);

                // browser may send xhr with port number only
                // to get next response.
                if (msg.size() == 0) {
                    if (sess == null) {
                        throw new RemoteException(
                                "XHR Session was not yet established.");
                    }
                    sendNextResponse(xhr, sess);
                    return;
                }
            }

            if (msg.size() < 1) {
                throw new RemoteException("Empty message on port:"+pid);
            }
            Port p = registry.getPort(pid);
            int h = msg.readI8() & 0xff;

            if (I) Log.info("reading "+(msg.size()+1)+" bytes from port:"+pid+
                    " head:"+h);

            // establish an xhr session if not found.
            if (xhr != null && sess == null) {
                sess = new XhrSession();
                sess.portId = pid;
                xhrSessions.add(sess);
            }

            switch (h >>> 6) {
            case Port.MESSAGE_RETURN:
                {
                    RemoteCall rc = p.removeRemoteCall();
                    Buffer buf2 = rc.buffer;
                    buf2.writeI8(h);
                    buf2.writeBuffer(msg);
                    rc.scan();
                    rc.isValid = true;
                    p.notifyRead();

                    if (xhr != null) {
                        sess.callDepth--;
                        // browser sent a return, which implies that
                        // initial xhr request was not yet finished.
                        if (sess.callDepth <= 0) {
                            throw new RemoteException(
                                    "XHR call depth mismatch");
                        }
                    }
                    break;
                }

            case Port.MESSAGE_SYNC_CALL:
                if (xhr != null) {
                    sess.callDepth++;
                }
                // fall through
            case Port.MESSAGE_ASYNC_CALL:
            case Port.MESSAGE_CHAINED_CALL:
                {
                    LocalCall lc = (LocalCall)New.get(LocalCall.class);
                    if (msg.size() < 4) {
                        throw new RemoteException("Too short header");
                    }
                    int oid = -msg.readI32();
                    if (D) Log.debug("oid:"+oid);
                    Buffer buf2 = lc.init(h, oid, registry);
                    buf2.writeBuffer(msg);
                    lc.scan();
                    p.addLocalCall(lc);
                    break;
                }

            default: return; // never happens
            }

            // close xhr request if complete.
            if (xhr != null) {
                // send dummy response, and close it.
                if (sess.callDepth == 0) {
                    xhrSessions.remove(sess);
                    try {
                        xhr.sendText(200, "");
                    } catch (IOException e) {
                        xhr.close(e);
                    }
                    return;
                }
                sendNextResponse(xhr, sess);
            }
        }
    }

    public void wait (Port p) {
        synchronized (this) {
            if (closed) {
                throw new RemoteException("Transport was closed");
            }
        }
        p.waitRead();
    }

    public void send (Port p, Buffer buf) {
        try {
            _send(p, buf);
        } catch (IOException e) {
            e.printStackTrace();
            throw new RemoteException("IO Error", e);
        }
    }

    private void _send (Port p, Buffer buf) throws IOException {

        buf.preWriteI32(p.id);

        synchronized (registry) {
            registry.checkDisposed();

            // if the port is attached with a xhr session
            // send through the xhr.
            XhrSession sess = getXhrSession(p.id);
            if (sess != null) {

                // trace call depth.
                // 0 means that there are no additional response via xhr.
                switch ((buf.peek(4) & 0xff) >> 6) {
                case Port.MESSAGE_SYNC_CALL:
                    sess.callDepth++;
                    break;
                case Port.MESSAGE_RETURN:
                    sess.callDepth--;
                    break;
                }

                // construct send buffer
                Buffer sb = new Buffer();
                int len = (buf.size() + 2) / 3 * 4;
                sb.writeRawString("HTTP/1.1 200 OK\r\n" +
                        "Content-Type: text/plain\r\n" +
                        "Content-Length: "+len+"\r\n" +
                        "Access-Control-Allow-Origin: *\r\n" +
                        "\r\n");
                sb.writeBase64(buf);

                if (sess == currentXhrSession && sess.response == null) {
                    // directly send the response
                    xmlHttpRequest.send(sb);
                    xmlHttpRequest.endResponse();
                    xmlHttpRequest = null;
                    currentXhrSession = null;

                    // close xhr session
                    if (sess.callDepth == 0) {
                        xhrSessions.remove(sess);
                    }
                } else {
                    sess.addResponse(sb);
                }
            } else { // via websocket
                if (webSocket != null) {
                    webSocket.sendMessage(buf);
                } else {
                    if (pendingMessages == null) {
                        pendingMessages = new ArrayList();
                    }
                    pendingMessages.add(new Buffer(buf));
                }
            }
        }

        buf.clear();
    }

    void close () {
        synchronized (this) {
            closed = true;
        }

        synchronized (registry) {
            if (webSocket != null) {
                webSocket.close("bound transport was closed");
                webSocket = null;
            }
            if (xmlHttpRequest != null) {
                xmlHttpRequest.close("bound transport was closed");
                xmlHttpRequest = null;
            }
            registry.dispose();
        }
    }

    private Buffer messageBuffer = new Buffer();

    public void handleHttpRequest (HttpReactor r) {
        // decode base64 content
        messageBuffer.clear();
        r.getContent().readBase64(messageBuffer);
        handleMessage(messageBuffer, r);
    }

    public void handleWebsocketOpen (HttpReactor r) {
    }

    public void handleWebsocketMessage (HttpReactor r) {
        handleMessage(r.getContent(), null);
    }

    public void handleWebsocketTextMessage (HttpReactor r) {
        messageBuffer.clear();
        r.getContent().readBase64(messageBuffer);
        handleMessage(messageBuffer, null);
    }

    public void handleWebsocketClose (HttpReactor r) {
        if (I) Log.info("closing transport:" + this);
        close ();
        httpServer.transportClosed(sessionId);
    }

    private static final long TIMEOUT = 60*1000;

    public long timeout;
    public void touchTimeout () {
        timeout = NativeIo.getMonotonicTime() + TIMEOUT;
    }
    public boolean isStale () {
        return (webSocket == null && (
                NativeIo.getMonotonicTime() - timeout > 0));
    }

    public String toString () {
        return "HttpTransport("+sessionId+")";
    }
}
